architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+# Attempt upgrade that encounters an issue
+> rad upgrade kubernetes --version v0.44.0
+
+Initiating Radius upgrade from v0.43.0 to v0.44.0...
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Compatible upgrade path
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✗ Applications Resource Provider (ERROR: Container image pull failed)
+
+ERROR: Upgrade failed during Applications Resource Provider update.
+Initiating automatic rollback to v0.43.0...
+  ✓ Restoring from backup (Not sure if this is needed)
+  ✓ Universal Control Plane reverted
+  ✓ Deployment Engine reverted
+  ✓ System verification complete
+  ✓ ...
+
+Rollback complete. System has been restored to v0.43.0.
+Review Kubernetes events and logs for more details on the failure.
+```
+
+**Result:**
+
+1. System detects failure during the upgrade process
+1. Automatic restore is initiated using the pre-upgrade backup
+1. All components are restored to their previous state
+1. User is informed of the failure and suggested next steps
+
+**Exceptions:**
+
+1. If the user data backup restoration fails (rare but possible)
+
+#### Scenario 4: Upgrading across multiple versions
+
+**User Story:** As a DevOps engineer, I need to upgrade Radius from an older version to the latest version in a single operation.
+
+**User Experience:**
+
+```bash
+# Check current version (significantly behind latest)
+> rad version
+RELEASE   VERSION   BICEP     COMMIT
+0.40.0    v0.40.0   0.31.93   1dd17270ec9bc9e725f314fa62c249406034edda
+
+# Upgrade directly to latest version
+> rad upgrade kubernetes --version latest
+
+Initiating Radius upgrade from v0.40.0 to v0.44.0 (latest stable)...
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Multiple version jump detected (v0.40.0 → v0.44.0)
+  ✓ Compatible upgrade path confirmed
+  ✓ Database schema changes detected
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✓ Applications Resource Provider
+  ✓ Controller
+Performing post-upgrade verification...
+  ✓ All components healthy
+  ✓ Database schema updated successfully
+
+Upgrade complete! Radius has been successfully upgraded to v0.44.0.
+Note: Your local Radius CLI is still v0.40.0. To upgrade your CLI, download the latest version.
+```
+
+**Result:**
+
+1. System detects a multi-version upgrade scenario
+1. Pre-flight checks validate that direct upgrade is supported
+1. All components are upgraded to the target version
+1. Verification confirms system health with the new version
+
+**Exceptions:**
+
+1. If direct upgrade path isn't supported between versions
+1. If database migrations encounter issues (this may be the case when we introduce Postgres as the data store)
+1. If intermediate upgrades are required first
+
+## Design
+
+### High-Level Design Diagram
+
+```mermaid
```

Comment:
This diagram appears to indicate that the user data backup runs in parallel to preflight checks but does not have to complete before the helm upgrade is applied. Is that the intended flow?

Reply:
No, that is not right. I will update it.